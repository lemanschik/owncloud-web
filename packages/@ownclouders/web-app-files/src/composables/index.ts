export * from './actions'
export * from './archiverService'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './selection'
export * from './shares'
export * from './spaces'
export * from './viewMode'
