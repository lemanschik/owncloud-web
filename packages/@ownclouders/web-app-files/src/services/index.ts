export * from './archiver'
export { default as Registry } from './registry'
