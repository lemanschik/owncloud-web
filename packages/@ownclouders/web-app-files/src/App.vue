<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div v-if="dragareaEnabled" class="dragarea" />
    <router-view tabindex="0" class="files-wrapper oc-width-expand oc-height-1-1 oc-flex-wrap" />
  </main>
</template>
<script lang="ts">
import { defineComponent, onBeforeUnmount, watch, ref } from 'vue'
import { useRoute, useStore } from 'web-pkg/src/composables'
import { eventBus } from 'web-pkg/src/services/eventBus'

export default defineComponent({
  setup() {
    const store = useStore<any>()
    const dragareaEnabled = ref(false)
    watch(useRoute(), () => {
      store.dispatch('Files/resetFileSelection')
    })

    const hideDropzone = () => {
      dragareaEnabled.value = false
    }
    const onDragOver = (event) => {
      dragareaEnabled.value = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }

    const dragOver = eventBus.subscribe('drag-over', onDragOver)
    const dragOut = eventBus.subscribe('drag-out', hideDropzone)
    const drop = eventBus.subscribe('drop', hideDropzone)

    onBeforeUnmount(() => {
      eventBus.unsubscribe('drag-over', dragOver)
      eventBus.unsubscribe('drag-out', dragOut)
      eventBus.unsubscribe('drop', drop)
    })
    return { dragareaEnabled }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-color-swatch-primary-muted);
}

#files {
  position: relative;
}
</style>
