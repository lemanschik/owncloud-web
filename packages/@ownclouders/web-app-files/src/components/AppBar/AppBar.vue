<template>
  <div id="files-app-bar" ref="filesAppBar" :class="{ 'files-app-bar-squashed': sideBarOpen }">
    <oc-hidden-announcer :announcement="selectedResourcesAnnouncement" level="polite" />
    <div class="files-topbar oc-py-s">
      <h1 class="oc-invisible-sr" v-text="pageTitle" />
      <div
        class="oc-flex oc-flex-middle files-app-bar-controls"
        :class="{
          'oc-flex-between': breadcrumbs.length || hasSharesNavigation,
          'oc-flex-right': !breadcrumbs.length && !hasSharesNavigation
        }"
      >
        <oc-breadcrumb
          v-if="showBreadcrumb"
          id="files-breadcrumb"
          data-testid="files-breadcrumbs"
          class="oc-flex oc-flex-middle"
          context-menu-padding="small"
          :show-context-actions="showContextActions"
          :items="breadcrumbs"
        >
          <template #contextMenu>
            <context-actions
              :action-options="{ space, resources: breadcrumbsContextActionsItems.filter(Boolean) }"
            />
          </template>
        </oc-breadcrumb>
        <portal-target v-if="showMobileNav" name="app.runtime.mobile.nav" />
        <shares-navigation v-if="hasSharesNavigation" />
        <div v-if="hasViewOptions || hasSidebarToggle" class="oc-flex">
          <view-options
            v-if="hasViewOptions"
            :view-modes="viewModes"
            :has-hidden-files="hasHiddenFiles"
            :has-file-extensions="hasFileExtensions"
            :has-pagination="hasPagination"
          />
          <sidebar-toggle v-if="hasSidebarToggle" :side-bar-open="sideBarOpen" />
        </div>
      </div>
      <div class="files-app-bar-actions oc-mt-xs">
        <div class="oc-flex-1 oc-flex oc-flex-start">
          <slot
            v-if="!selectedFiles.length || (showActionsOnSelection && selectedFiles.length === 1)"
            name="actions"
            :limited-screen-space="limitedScreenSpace"
          />
          <batch-actions
            v-if="showBatchActions"
            :actions="batchActions"
            :action-options="{ space, resources: selectedFiles }"
            :limited-screen-space="limitedScreenSpace"
          />
        </div>
      </div>
      <slot name="content" />
    </div>
  </div>
</template>

<script lang="ts">
import last from 'lodash-es/last'
import { computed, defineComponent, inject, PropType, Ref, unref } from 'vue'
import { mapGetters, mapState, mapMutations } from 'vuex'
import { Resource } from 'web-client'
import {
  isPersonalSpaceResource,
  isProjectSpaceResource,
  SpaceResource
} from 'web-client/src/helpers'
import BatchActions from 'web-pkg/src/components/BatchActions.vue'
import { isLocationTrashActive } from '../../router'
import ContextActions from '../FilesList/ContextActions.vue'
import SharesNavigation from './SharesNavigation.vue'
import SidebarToggle from './SidebarToggle.vue'
import ViewOptions from './ViewOptions.vue'
import { ViewMode } from 'web-pkg/src/ui/types'
import {
  useFileActionsAcceptShare,
  useFileActionsClearSelection,
  useFileActionsCopy,
  useFileActionsDeclineShare,
  useFileActionsDelete,
  useFileActionsDownloadArchive,
  useFileActionsDownloadFile,
  useFileActionsEmptyTrashBin,
  useFileActionsMove,
  useFileActionsRestore
} from 'web-app-files/src/composables/actions'
import { useStore } from 'web-pkg/src'
import { BreadcrumbItem } from 'design-system/src/components/OcBreadcrumb/types'
import { useActiveLocation } from 'web-app-files/src/composables'

export default defineComponent({
  components: {
    BatchActions,
    ContextActions,
    SharesNavigation,
    SidebarToggle,
    ViewOptions
  },
  props: {
    breadcrumbs: {
      type: Array as PropType<BreadcrumbItem[]>,
      default: () => []
    },
    breadcrumbsContextActionsItems: {
      type: Array as PropType<Resource[]>,
      default: () => []
    },
    viewModes: {
      type: Array as PropType<ViewMode[]>,
      default: () => []
    },
    hasBulkActions: { type: Boolean, default: false },
    hasSharesNavigation: { type: Boolean, default: false },
    hasSidebarToggle: { type: Boolean, default: true },
    hasViewOptions: { type: Boolean, default: true },
    hasHiddenFiles: { type: Boolean, default: true },
    hasFileExtensions: { type: Boolean, default: true },
    hasPagination: { type: Boolean, default: true },
    showActionsOnSelection: { type: Boolean, default: false },
    sideBarOpen: { type: Boolean, default: false },
    space: {
      type: Object as PropType<SpaceResource>,
      required: false,
      default: null
    }
  },
  setup(props) {
    const store = useStore()

    const { actions: acceptShareActions } = useFileActionsAcceptShare({ store })
    const { actions: clearSelectionActions } = useFileActionsClearSelection({ store })
    const { actions: copyActions } = useFileActionsCopy({ store })
    const { actions: declineShareActions } = useFileActionsDeclineShare({ store })
    const { actions: deleteActions } = useFileActionsDelete({ store })
    const { actions: downloadArchiveActions } = useFileActionsDownloadArchive({ store })
    const { actions: downloadFileActions } = useFileActionsDownloadFile()
    const { actions: emptyTrashBinActions } = useFileActionsEmptyTrashBin({ store })
    const { actions: moveActions } = useFileActionsMove({ store })
    const { actions: restoreActions } = useFileActionsRestore({ store })

    const batchActions = computed(() => {
      return [
        ...unref(clearSelectionActions),
        ...unref(acceptShareActions),
        ...unref(declineShareActions),
        ...unref(downloadArchiveActions),
        ...unref(downloadFileActions),
        ...unref(moveActions),
        ...unref(copyActions),
        ...unref(emptyTrashBinActions),
        ...unref(deleteActions),
        ...unref(restoreActions)
      ].filter((item) =>
        item.isEnabled({ space: props.space, resources: store.getters['Files/selectedFiles'] })
      )
    })

    const spaces = computed<SpaceResource[]>(() =>
      store.getters['runtime/spaces/spaces'].filter(
        (s) => isPersonalSpaceResource(s) || isProjectSpaceResource(s)
      )
    )

    const isMobileWidth = inject<Ref<boolean>>('isMobileWidth')
    const isTrashLocation = useActiveLocation(isLocationTrashActive, 'files-trash-generic')
    const showBreadcrumb = computed(() => {
      if (!unref(isMobileWidth) && props.breadcrumbs.length) {
        return true
      }
      if (unref(isTrashLocation) && unref(spaces).length === 1) {
        return false
      }
      return props.breadcrumbs.length > 1
    })
    const showMobileNav = computed(() => {
      if (unref(isTrashLocation) && unref(spaces).length === 1) {
        return props.breadcrumbs.length <= 2
      }
      return props.breadcrumbs.length <= 1
    })

    return {
      batchActions,
      showBreadcrumb,
      showMobileNav
    }
  },
  data: function () {
    return {
      resizeObserver: new ResizeObserver(this.onResize as ResizeObserverCallback),
      limitedScreenSpace: false
    }
  },
  computed: {
    ...mapGetters('Files', ['files', 'selectedFiles']),
    ...mapState('Files', ['areHiddenFilesShown', 'areFileExtensionsShown']),

    pageTitle() {
      const title = this.$route.meta.title || ''
      return this.$gettext(title)
    },
    showContextActions() {
      return last<BreadcrumbItem>(this.breadcrumbs).allowContextActions
    },
    showBatchActions() {
      return (
        this.hasBulkActions &&
        (this.selectedFiles.length > 1 ||
          isLocationTrashActive(this.$router, 'files-trash-generic'))
      )
    },
    selectedResourcesAnnouncement() {
      if (this.selectedFiles.length === 0) {
        return this.$gettext('No items selected.')
      }
      const translated = this.$ngettext(
        '%{ amount } item selected. Actions are available above the table.',
        '%{ amount } items selected. Actions are available above the table.',
        this.selectedFiles.length
      )
      return this.$gettextInterpolate(translated, { amount: this.selectedFiles.length })
    }
  },
  mounted() {
    this.resizeObserver.observe(this.$refs.filesAppBar as HTMLElement)
  },
  beforeUnmount() {
    this.resizeObserver.unobserve(this.$refs.filesAppBar as HTMLElement)
  },

  created() {
    // Storage returns a string so we need to convert it into a boolean
    const areHiddenFilesShown = window.localStorage.getItem('oc_hiddenFilesShown') || 'false'
    const areHiddenFilesShownBoolean = areHiddenFilesShown === 'true'

    if (areHiddenFilesShownBoolean !== this.areHiddenFilesShown) {
      this.SET_HIDDEN_FILES_VISIBILITY(areHiddenFilesShownBoolean)
    }

    // Storage returns a string so we need to convert it into a boolean
    const areFileExtensionsShown = window.localStorage.getItem('oc_fileExtensionsShown') || 'true'
    const areFileExtensionsShownBoolean = areFileExtensionsShown === 'true'

    if (areFileExtensionsShownBoolean !== this.areFileExtensionsShown) {
      this.SET_FILE_EXTENSIONS_VISIBILITY(areFileExtensionsShownBoolean)
    }
  },

  methods: {
    ...mapMutations('Files', ['SET_HIDDEN_FILES_VISIBILITY', 'SET_FILE_EXTENSIONS_VISIBILITY']),

    onResize() {
      this.limitedScreenSpace = this.sideBarOpen
        ? window.innerWidth <= 1280
        : window.innerWidth <= 1000
    }
  }
})
</script>

<style lang="scss" scoped>
#files-app-bar {
  background-color: var(--oc-color-background-default);
  border-top-right-radius: 15px;
  box-sizing: border-box;
  z-index: 2;
  position: sticky;
  padding: 0 var(--oc-space-medium);
  top: 0;

  .files-app-bar-controls {
    min-height: 52px;

    @media (max-width: $oc-breakpoint-xsmall-max) {
      justify-content: space-between;
    }
  }

  .files-app-bar-actions {
    align-items: center;
    display: flex;
    gap: var(--oc-space-small);
    justify-content: flex-end;
    min-height: 3rem;
  }

  #files-breadcrumb {
    min-height: 2rem;
  }
}
</style>
