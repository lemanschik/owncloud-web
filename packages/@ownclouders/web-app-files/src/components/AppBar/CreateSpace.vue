<template>
  <oc-button
    id="new-space-menu-btn"
    key="new-space-menu-btn-enabled"
    v-oc-tooltip="$gettext('Create a new space')"
    :aria-label="$gettext('Create a new space')"
    appearance="filled"
    variation="primary"
    @click="showCreateSpaceModal"
  >
    <oc-icon name="add" />
    <span v-translate>New Space</span>
  </oc-button>
</template>

<script lang="ts">
import { mapActions, mapMutations } from 'vuex'
import { defineComponent } from 'vue'
import { useClientService } from 'web-pkg/src/composables'
import { useCreateSpace } from 'web-app-files/src/composables'
import { useSpaceHelpers } from 'web-pkg/src/composables/spaces'

export default defineComponent({
  setup() {
    const clientService = useClientService()
    const { createSpace } = useCreateSpace()
    const { checkSpaceNameModalInput } = useSpaceHelpers()
    return { clientService, createSpace, checkSpaceNameModalInput }
  },
  methods: {
    ...mapActions(['showMessage', 'createModal', 'hideModal', 'setModalInputErrorMessage']),
    ...mapMutations('runtime/spaces', ['UPSERT_SPACE']),
    ...mapMutations('Files', ['UPSERT_RESOURCE', 'UPDATE_RESOURCE_FIELD']),

    showCreateSpaceModal() {
      const modal = {
        variation: 'passive',
        title: this.$gettext('Create a new space'),
        cancelText: this.$gettext('Cancel'),
        confirmText: this.$gettext('Create'),
        hasInput: true,
        inputLabel: this.$gettext('Space name'),
        inputValue: this.$gettext('New space'),
        onCancel: this.hideModal,
        onConfirm: this.addNewSpace,
        onInput: this.checkSpaceNameModalInput
      }

      this.createModal(modal)
    },

    async addNewSpace(name) {
      try {
        const createdSpace = await this.createSpace(name)
        this.hideModal()
        this.UPSERT_RESOURCE(createdSpace)
        this.UPSERT_SPACE(createdSpace)
      } catch (error) {
        console.error(error)
        this.showMessage({
          title: this.$gettext('Creating space failedâ€¦'),
          status: 'danger'
        })
      }
    }
  }
})
</script>
