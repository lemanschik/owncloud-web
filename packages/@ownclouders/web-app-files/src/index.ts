import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceDriveResolver from './views/spaces/DriveResolver.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import TrashOverview from './views/trash/Overview.vue'
import translations from '../l10n/translations.json'
import quickActions from './quickActions'
import store from './store'
import { SDKSearch } from './search'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { Registry, ArchiverService } from './services'
import fileSideBars from './fileSideBars'
import { buildRoutes } from './router'
import get from 'lodash-es/get'
import { AppNavigationItem, AppReadyHookArgs } from 'web-pkg/src/apps'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'
import { isPersonalSpaceResource } from 'web-client/src/helpers'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
const navItems = (context): AppNavigationItem[] => {
  return [
    {
      name(capabilities) {
        return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
      },
      icon: appInfo.icon,
      route: {
        path: `/${appInfo.id}/spaces/personal`
      },
      enabled(capabilities) {
        if (!capabilities.spaces?.enabled) {
          return true
        }
        return !!context?.$store?.getters['runtime/spaces/spaces'].find((drive) =>
          isPersonalSpaceResource(drive)
        )
      }
    },
    {
      name: $gettext('Favorites'),
      icon: 'star',
      route: {
        path: `/${appInfo.id}/favorites`
      },
      enabled(capabilities) {
        return capabilities.files?.favorites
      }
    },
    {
      name: $gettext('Shares'),
      icon: 'share-forward',
      route: {
        path: `/${appInfo.id}/shares`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/share` }],
      enabled(capabilities) {
        return capabilities.files_sharing?.api_enabled !== false
      }
    },
    {
      name: $gettext('Spaces'),
      icon: 'layout-grid',
      route: {
        path: `/${appInfo.id}/spaces/projects`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/project` }],
      enabled(capabilities) {
        return capabilities.spaces?.projects
      }
    },
    {
      name: $gettext('Deleted files'),
      icon: 'delete-bin-5',
      route: {
        path: `/${appInfo.id}/trash/overview`
      },
      activeFor: [{ path: `/${appInfo.id}/trash` }],
      enabled(capabilities) {
        return capabilities.dav?.trashbin === '1.0'
      }
    }
  ]
}

export default {
  appInfo,
  store,
  routes: buildRoutes({
    App,
    Favorites,
    FilesDrop,
    SearchResults,
    Shares: {
      SharedViaLink,
      SharedWithMe,
      SharedWithOthers
    },
    Spaces: {
      DriveResolver: SpaceDriveResolver,
      Projects: SpaceProjects
    },
    Trash: {
      Overview: TrashOverview
    }
  }),
  navItems,
  quickActions,
  translations,
  ready({ router, store, globalProperties }: AppReadyHookArgs) {
    const { $clientService } = globalProperties
    Registry.sdkSearch = new SDKSearch(store, router, $clientService)

    // when discussing the boot process of applications we need to implement a
    // registry that does not rely on call order, aka first register "on" and only after emit.
    eventBus.publish('app.search.register.provider', Registry.sdkSearch)
    globalProperties.$archiverService = new ArchiverService(
      $clientService,
      store.getters.configuration.server || window.location.origin,
      get(store, 'getters.capabilities.files.archivers', [
        {
          enabled: true,
          version: '1.0.0',
          formats: ['tar', 'zip'],
          archiver_url: `${store.getters.configuration.server}index.php/apps/files/ajax/download.php`
        }
      ])
    )
  }
}
