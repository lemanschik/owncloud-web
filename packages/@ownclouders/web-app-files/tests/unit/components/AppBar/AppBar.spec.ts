import { ref } from 'vue'
import AppBar from 'web-app-files/src/components/AppBar/AppBar.vue'
import { mock, mockDeep } from 'jest-mock-extended'
import { Resource, SpaceResource } from 'web-client'
import {
  createStore,
  defaultComponentMocks,
  defaultPlugins,
  shallowMount,
  defaultStoreMockOptions,
  RouteLocation
} from 'web-test-helpers'
import { ViewModeConstants } from 'web-app-files/src/composables'
import { ArchiverService } from 'web-app-files/src/services'

const selectors = {
  ocBreadcrumbStub: 'oc-breadcrumb-stub',
  batchActionsStub: 'batch-actions-stub',
  sharesNavigationStub: 'shares-navigation-stub',
  viewOptionsStub: 'view-options-stub',
  sidebarToggleStub: 'sidebar-toggle-stub',
  mobileNavPortal: 'portal-target[name="app.runtime.mobile.nav"]'
}

const selectedFiles = [mockDeep<Resource>(), mockDeep<Resource>()]
const actionSlot = "<button class='action-slot'>Click</button>"
const contentSlot = "<div class='content-slot'>Foo</div>"

const breadcrumbItems = [
  { text: 'Example1', to: '/' },
  { text: 'Example2', to: '/foo' }
]
const breadCrumbItemWithContextActionAllowed = {
  text: 'Example Special',
  to: '/bar',
  allowContextActions: true
}

describe('AppBar component', () => {
  describe('renders', () => {
    it('by default no breadcrumbs, no bulkactions, no sharesnavigation but viewoptions and sidebartoggle', () => {
      const { wrapper } = getShallowWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
    describe('breadcrumbs', () => {
      it('if given, by default without breadcrumbsContextActionsItems', () => {
        const { wrapper } = getShallowWrapper([], {}, { breadcrumbs: breadcrumbItems })
        expect(wrapper.find(selectors.ocBreadcrumbStub).exists()).toBeTruthy()
        expect(wrapper.findComponent<any>(selectors.ocBreadcrumbStub).props('items')).toEqual(
          breadcrumbItems
        )
      })
      it('if given, with breadcrumbsContextActionsItems if allowed on last breadcrumb item', () => {
        const { wrapper } = getShallowWrapper(
          [],
          {},
          { breadcrumbs: [...breadcrumbItems, breadCrumbItemWithContextActionAllowed] }
        )
        expect(wrapper.find(selectors.ocBreadcrumbStub).exists()).toBeTruthy()
        expect(wrapper.findComponent<any>(selectors.ocBreadcrumbStub).props('items')).toEqual([
          ...breadcrumbItems,
          breadCrumbItemWithContextActionAllowed
        ])
      })
      it('not if no breadcrumb items given', () => {
        const { wrapper } = getShallowWrapper([], {}, { breadcrumbs: [] })
        expect(wrapper.find(selectors.ocBreadcrumbStub).exists()).toBeFalsy()
      })
      it('not if one breadcrumb item is given in mobile view', () => {
        const { wrapper } = getShallowWrapper(
          [],
          {},
          { breadcrumbs: [breadcrumbItems[0]] },
          mock<RouteLocation>(),
          true
        )
        expect(wrapper.find(selectors.ocBreadcrumbStub).exists()).toBeFalsy()
      })
    })
    describe('bulkActions', () => {
      it('if enabled', () => {
        const { wrapper } = getShallowWrapper(selectedFiles, {}, { hasBulkActions: true })
        expect(wrapper.find(selectors.batchActionsStub).exists()).toBeTruthy()
      })
      it('if 1 file selected on trash routes', () => {
        const { wrapper } = getShallowWrapper(
          [selectedFiles[0]],
          {},
          { hasBulkActions: true },
          mock<RouteLocation>({
            name: 'files-trash-generic',
            path: '/files/trash/personal/admin'
          })
        )
        expect(wrapper.find(selectors.batchActionsStub).exists()).toBeTruthy()
      })
      it('not if 1 file selected', () => {
        const { wrapper } = getShallowWrapper([selectedFiles[0]], {}, { hasBulkActions: true })
        expect(wrapper.find(selectors.batchActionsStub).exists()).toBeFalsy()
      })
    })
    describe('mobile navigation portal', () => {
      it.each([
        { items: [], shows: true },
        { items: [breadcrumbItems[0]], shows: true },
        { items: [breadcrumbItems[0], breadcrumbItems[1]], shows: false }
      ])('if less than 2 breadcrumb items given', ({ items, shows }) => {
        const { wrapper } = getShallowWrapper([], {}, { breadcrumbs: items })
        expect(wrapper.find(selectors.mobileNavPortal).exists()).toBe(shows)
      })
    })
    describe('sharesNavigation', () => {
      it('if enabled', () => {
        const { wrapper } = getShallowWrapper([], {}, { hasSharesNavigation: true })
        expect(wrapper.find(selectors.sharesNavigationStub).exists()).toBeTruthy()
      })
    })
    describe('viewoptions and sidebartoggle', () => {
      it('only viewoptions if sidebartoggle is disabled', () => {
        const { wrapper } = getShallowWrapper([], {}, { hasSidebarToggle: false })
        expect(wrapper.find(selectors.viewOptionsStub).exists()).toBeTruthy()
        expect(wrapper.find(selectors.sidebarToggleStub).exists()).toBeFalsy()
      })
      it('only sidebartoggle if viewoptions is disabled', () => {
        const { wrapper } = getShallowWrapper([], {}, { hasViewOptions: false })
        expect(wrapper.find(selectors.viewOptionsStub).exists()).toBeFalsy()
        expect(wrapper.find(selectors.sidebarToggleStub).exists()).toBeTruthy()
      })
      it('neither if both are disabled', () => {
        const { wrapper } = getShallowWrapper(
          [],
          {},
          { hasSidebarToggle: false, hasViewOptions: false }
        )
        expect(wrapper.find(selectors.viewOptionsStub).exists()).toBeFalsy()
        expect(wrapper.find(selectors.sidebarToggleStub).exists()).toBeFalsy()
      })
      it('passes viewModes array to ViewOptions', () => {
        const viewModes = [ViewModeConstants.tilesView]
        const { wrapper } = getShallowWrapper([], {}, { hasViewOptions: true, viewModes })
        expect(wrapper.findComponent<any>(selectors.viewOptionsStub).props('viewModes')).toEqual(
          viewModes
        )
      })
    })
    it('if given, with content in the actions slot', () => {
      const { wrapper } = getShallowWrapper([], { actions: actionSlot })
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('if given, with content in the content slot', () => {
      const { wrapper } = getShallowWrapper([], { content: contentSlot })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})

function getShallowWrapper(
  selected = [],
  slots = {},
  props: { [key: string]: any } = {
    breadcrumbs: [],
    displayViewModeSwitch: false,
    hasBulkActions: false,
    hasSharesNavigation: false,
    hasSidebarToggle: true,
    hasViewOptions: true
  },
  currentRoute = mock<RouteLocation>({
    name: 'files-spaces-generic',
    path: '/files/spaces/personal/admin'
  }),
  isMobileWidth = false
) {
  const mocks = {
    ...defaultComponentMocks({
      currentRoute
    }),
    $archiverService: mock<ArchiverService>()
  }
  mocks.$route.meta.title = 'ExampleTitle'
  const storeOptions = defaultStoreMockOptions
  storeOptions.modules.Files.getters.selectedFiles.mockImplementation(() => selected)
  const store = createStore(storeOptions)
  return {
    wrapper: shallowMount(AppBar, {
      props: { ...props, space: mock<SpaceResource>() },
      slots,
      global: {
        plugins: [...defaultPlugins(), store],
        provide: { isMobileWidth: ref(isMobileWidth) },
        mocks
      }
    })
  }
}
