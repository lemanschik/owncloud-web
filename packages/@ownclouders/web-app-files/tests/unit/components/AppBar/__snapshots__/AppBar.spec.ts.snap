// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar component renders by default no breadcrumbs, no bulkactions, no sharesnavigation but viewoptions and sidebartoggle 1`] = `
<div class="" displayviewmodeswitch="false" id="files-app-bar">
  <oc-hidden-announcer-stub announcement="No items selected." level="polite"></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-middle files-app-bar-controls oc-flex-right">
      <!--v-if-->
      <portal-target name="app.runtime.mobile.nav"></portal-target>
      <!--v-if-->
      <div class="oc-flex">
        <view-options-stub hasfileextensions="true" hashiddenfiles="true" haspagination="true" viewmodes=""></view-options-stub>
        <sidebar-toggle-stub sidebaropen="false"></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions oc-mt-xs">
      <div class="oc-flex-1 oc-flex oc-flex-start">
        <!--v-if-->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component renders if given, with content in the actions slot 1`] = `
<div class="" displayviewmodeswitch="false" id="files-app-bar">
  <oc-hidden-announcer-stub announcement="No items selected." level="polite"></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-middle files-app-bar-controls oc-flex-right">
      <!--v-if-->
      <portal-target name="app.runtime.mobile.nav"></portal-target>
      <!--v-if-->
      <div class="oc-flex">
        <view-options-stub hasfileextensions="true" hashiddenfiles="true" haspagination="true" viewmodes=""></view-options-stub>
        <sidebar-toggle-stub sidebaropen="false"></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions oc-mt-xs">
      <div class="oc-flex-1 oc-flex oc-flex-start">
        <button class="action-slot">Click</button>
        <!--v-if-->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component renders if given, with content in the content slot 1`] = `
<div class="" displayviewmodeswitch="false" id="files-app-bar">
  <oc-hidden-announcer-stub announcement="No items selected." level="polite"></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-middle files-app-bar-controls oc-flex-right">
      <!--v-if-->
      <portal-target name="app.runtime.mobile.nav"></portal-target>
      <!--v-if-->
      <div class="oc-flex">
        <view-options-stub hasfileextensions="true" hashiddenfiles="true" haspagination="true" viewmodes=""></view-options-stub>
        <sidebar-toggle-stub sidebaropen="false"></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions oc-mt-xs">
      <div class="oc-flex-1 oc-flex oc-flex-start">
        <!--v-if-->
      </div>
    </div>
    <div class="content-slot">Foo</div>
  </div>
</div>
`;
